[JSC] Add dragonbox for double-conversion

hyjorc1 opened the pull request:
#### 85e669f3c3e8d80e43a55e4c1c669bb2dd63bf33
<pre>
[JSC] Add dragonbox for double-conversion
<a href="https://bugs.webkit.org/show_bug.cgi?id=264284">https://bugs.webkit.org/show_bug.cgi?id=264284</a>
<a href="https://rdar.apple.com/118018415">rdar://118018415</a>

Reviewed by Yusuke Suzuki.

Dragonbox[1] is a float-to-string conversion algorithm based on a
algorithm Schubfach and inspired by Grisu and Grisu-Exact. The new
conversion brings us JetStream progressions. Currently, the new
conversion only supports ToExponential and ToShortest. This is
because the precision specification is incompatible with finding
the shortest representation (Dragonbox), which requires a completely
different algorithm[2]. The author is working on another repository[3] with more
complete feature set including the precision mode, which seems profitable
and competitive to current double-conversion (tested locally and the results
are shown below).

FP Precision vs double-conversion Precision:

    fPToExponential               -&gt; 0.001016 sec with precision 0
    doubleConversionToExponential -&gt; 0.000950 sec with precision 0
    fPToExponential               -&gt; 0.001358 sec with precision 1
    doubleConversionToExponential -&gt; 0.001199 sec with precision 1
    fPToExponential               -&gt; 0.001309 sec with precision 2
    doubleConversionToExponential -&gt; 0.001226 sec with precision 2
    fPToExponential               -&gt; 0.001135 sec with precision 3  (faster)
    doubleConversionToExponential -&gt; 0.001310 sec with precision 3
    ...
    fPToExponential               -&gt; 0.002514 sec with precision 48 (~15 times faster)
    doubleConversionToExponential -&gt; 0.037997 sec with precision 48
    ...
    fPToExponential               -&gt; 0.002511 sec with precision 99 (~18 times faster)
    doubleConversionToExponential -&gt; 0.045865 sec with precision 99

The overhead for FP is due to the computation of the digits segments for
significand since each segment has fixed size of 9. For example, given a
double

    0.398735806489994737411564074136549606919288635253906250000000

FP&apos;s algorithm will compute and iterate each significand segment one by one
with fixed size 9.

    1st significand segment = 398735806
    2nd significand segment = 489994737
    ...

So, even the precision arg is less than 9, the FP would compute the 1st
significand segment with size 9. And this is the major overhead cost
for current FP with small precision arg. On the other side, for double conversion
implemented in our codebase, it computes each significand digit one by one.
And that explains why current FP is slightly slower when precision arg is small
but much faster when precision arg is large.

[1] <a href="https://github.com/jk-jeon/dragonbox">https://github.com/jk-jeon/dragonbox</a>
[2] <a href="https://github.com/jk-jeon/dragonbox/issues/38">https://github.com/jk-jeon/dragonbox/issues/38</a>
[3] <a href="https://github.com/jk-jeon/fp">https://github.com/jk-jeon/fp</a>

* Source/JavaScriptCore/runtime/NumberPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/dragonbox/detail/bits.h: Added.
* Source/WTF/wtf/dragonbox/detail/cache_holder.h: Added.
* Source/WTF/wtf/dragonbox/detail/decimal_fp.h: Added.
* Source/WTF/wtf/dragonbox/detail/div.h: Added.
* Source/WTF/wtf/dragonbox/detail/log.h: Added.
* Source/WTF/wtf/dragonbox/detail/policy.h: Added.
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_to_even::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_to_odd::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_toward_plus_infinity::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_toward_minus_infinity::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_toward_zero::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_away_from_zero::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::detail::nearest_always_closed::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::detail::nearest_always_open::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_to_even_static_boundary::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_to_odd_static_boundary::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_toward_plus_infinity_static_boundary::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_toward_minus_infinity_static_boundary::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::toward_plus_infinity::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::toward_minus_infinity::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::toward_zero::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::away_from_zero::decltype):
* Source/WTF/wtf/dragonbox/detail/policy_holder.h: Added.
(WTF::dragonbox::detail::policy_impl::check_policy_validity):
(WTF::dragonbox::detail::policy_impl::check_policy_list_validity):
(WTF::dragonbox::detail::policy_impl::make_policy_holder):
(WTF::dragonbox::detail::decltype):
* Source/WTF/wtf/dragonbox/detail/util.h: Added.
* Source/WTF/wtf/dragonbox/detail/wuint.h: Added.
* Source/WTF/wtf/dragonbox/dragonbox.h: Added.
(WTF::dragonbox::to_exponential_max_string_length):
(WTF::dragonbox::to_shortest_max_string_length):
(WTF::dragonbox::max_string_length):
(WTF::dragonbox::valid_shortest_representation):
(WTF::dragonbox::count_digits_base10_with_max_17):
(WTF::dragonbox::compute_power_with_max_16):
* Source/WTF/wtf/dragonbox/dragonbox_to_chars.cpp: Added.
* Source/WTF/wtf/dragonbox/dragonbox_to_chars.h: Added.
(WTF::dragonbox::ToExponential):
(WTF::dragonbox::ToShortest):
* Source/WTF/wtf/dragonbox/ieee754_format.h: Added.
* Source/WTF/wtf/dtoa.cpp:
(WTF::numberToString):
* Source/WTF/wtf/dtoa.h:
* Source/WTF/wtf/dtoa/double-conversion.cc:
* Source/WTF/wtf/dtoa/utils.h:
(WTF::double_conversion::validShortestRepresentation):
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/DragonBoxTest.cpp: Added.
(TestWebKitAPI::DragonBoxTest::DragonBoxTest):
(TestWebKitAPI::DragonBoxTest::randomFloat):
(TestWebKitAPI::DragonBoxTest::randomDouble):
(TestWebKitAPI::DragonBoxTest::dragonBoxToString):
(TestWebKitAPI::DragonBoxTest::dragonBoxToExponential):
(TestWebKitAPI::DragonBoxTest::doubleConversionToString):
(TestWebKitAPI::DragonBoxTest::doubleConversionToExponential):
(TestWebKitAPI::TEST):

Canonical link: <a href="https://commits.webkit.org/271612@main">https://commits.webkit.org/271612@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/2d8834fcaf2b76cd1e13bc15cd40c6f3c3923edb

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/29025 "120 style errors") | [  ~~🛠 ios~~](https://ews-build.webkit.org/#/builders/48/builds/7700 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30380 "Built successfully") | [  ~~🛠 wpe~~](https://ews-build.webkit.org/#/builders/5/builds/31640 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 wincairo~~](https://ews-build.webkit.org/#/builders/32/builds/26441 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/29616 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9830 "Built successfully") | [  ~~🛠 mac-AS-debug~~](https://ews-build.webkit.org/#/builders/51/builds/5013 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/26457 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/29297 "Passed tests") | [⏳ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 api-mac~~](https://ews-build.webkit.org/#/builders/18/builds/24898 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-wpe~~](https://ews-build.webkit.org/#/builders/41/builds/5520 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [⏳ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/iOS-17-Simulator-WPT-WK2-Tests-EWS "Waiting to run tests") | [  ~~🧪 mac-wk1~~](https://ews-build.webkit.org/#/builders/10/builds/25925 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 gtk~~](https://ews-build.webkit.org/#/builders/2/builds/32977 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc~~](https://ews-build.webkit.org/#/builders/20/builds/25054 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-ios~~](https://ews-build.webkit.org/#/builders/13/builds/26523 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-wk2~~](https://ews-build.webkit.org/#/builders/36/builds/26361 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/31896 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [  ~~🛠 🧪 jsc-arm64~~](https://ews-build.webkit.org/#/builders/12/builds/29373 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🛠 tv~~](https://ews-build.webkit.org/#/builders/44/builds/5626 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 mac-AS-debug-wk2~~](https://ews-build.webkit.org/#/builders/50/builds/3803 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [  ~~🧪 api-gtk~~](https://ews-build.webkit.org/#/builders/21/builds/29679 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| | [  ~~🛠 tv-sim~~](https://ews-build.webkit.org/#/builders/46/builds/7283 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/35638 "Failed to checkout and rebase branch from PR 20467") | 
| | [  ~~🛠 watch~~](https://ews-build.webkit.org/#/builders/43/builds/6124 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | [❌ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/35/builds/35638 "Failed to checkout and rebase branch from PR 20467") | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/3731 "Built successfully and passed tests") | [  ~~🛠 watch-sim~~](https://ews-build.webkit.org/#/builders/45/builds/6147 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | | | 
<!--EWS-Status-Bubble-End-->

Constellation reviewed (approved):
Put LICENSE-dragonbox.txt in `Source/WTF` dir (as the same to other LICENSE-LLVM.txt etc.).

hyjorc1 committed:
[JSC] Add dragonbox for double-conversion
https://bugs.webkit.org/show_bug.cgi?id=264284
rdar://118018415

Reviewed by Yusuke Suzuki.

Dragonbox[1] is a float-to-string conversion algorithm based on a
algorithm Schubfach and inspired by Grisu and Grisu-Exact. The new
conversion brings us JetStream progressions. Currently, the new
conversion only supports ToExponential and ToShortest. This is
because the precision specification is incompatible with finding
the shortest representation (Dragonbox), which requires a completely
different algorithm[2]. The author is working on another repository[3] with more
complete feature set including the precision mode, which seems profitable
and competitive to current double-conversion (tested locally and the results
are shown below).

FP Precision vs double-conversion Precision:

    fPToExponential               -> 0.001016 sec with precision 0
    doubleConversionToExponential -> 0.000950 sec with precision 0
    fPToExponential               -> 0.001358 sec with precision 1
    doubleConversionToExponential -> 0.001199 sec with precision 1
    fPToExponential               -> 0.001309 sec with precision 2
    doubleConversionToExponential -> 0.001226 sec with precision 2
    fPToExponential               -> 0.001135 sec with precision 3  (faster)
    doubleConversionToExponential -> 0.001310 sec with precision 3
    ...
    fPToExponential               -> 0.002514 sec with precision 48 (~15 times faster)
    doubleConversionToExponential -> 0.037997 sec with precision 48
    ...
    fPToExponential               -> 0.002511 sec with precision 99 (~18 times faster)
    doubleConversionToExponential -> 0.045865 sec with precision 99

The overhead for FP is due to the computation of the digits segments for
significand since each segment has fixed size of 9. For example, given a
double

    0.398735806489994737411564074136549606919288635253906250000000

FP's algorithm will compute and iterate each significand segment one by one
with fixed size 9.

    1st significand segment = 398735806
    2nd significand segment = 489994737
    ...

So, even the precision arg is less than 9, the FP would compute the 1st
significand segment with size 9. And this is the major overhead cost
for current FP with small precision arg. On the other side, for double conversion
implemented in our codebase, it computes each significand digit one by one.
And that explains why current FP is slightly slower when precision arg is small
but much faster when precision arg is large.

[1] https://github.com/jk-jeon/dragonbox
[2] https://github.com/jk-jeon/dragonbox/issues/38
[3] https://github.com/jk-jeon/fp

* Source/JavaScriptCore/runtime/NumberPrototype.cpp:
(JSC::JSC_DEFINE_HOST_FUNCTION):
* Source/WTF/WTF.xcodeproj/project.pbxproj:
* Source/WTF/wtf/CMakeLists.txt:
* Source/WTF/wtf/dragonbox/detail/bits.h: Added.
* Source/WTF/wtf/dragonbox/detail/cache_holder.h: Added.
* Source/WTF/wtf/dragonbox/detail/decimal_fp.h: Added.
* Source/WTF/wtf/dragonbox/detail/div.h: Added.
* Source/WTF/wtf/dragonbox/detail/log.h: Added.
* Source/WTF/wtf/dragonbox/detail/policy.h: Added.
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_to_even::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_to_odd::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_toward_plus_infinity::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_toward_minus_infinity::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_toward_zero::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_away_from_zero::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::detail::nearest_always_closed::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::detail::nearest_always_open::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_to_even_static_boundary::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_to_odd_static_boundary::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_toward_plus_infinity_static_boundary::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::nearest_toward_minus_infinity_static_boundary::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::toward_plus_infinity::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::toward_minus_infinity::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::toward_zero::decltype):
(WTF::dragonbox::detail::policy_impl::decimal_to_binary_rounding::away_from_zero::decltype):
* Source/WTF/wtf/dragonbox/detail/policy_holder.h: Added.
(WTF::dragonbox::detail::policy_impl::check_policy_validity):
(WTF::dragonbox::detail::policy_impl::check_policy_list_validity):
(WTF::dragonbox::detail::policy_impl::make_policy_holder):
(WTF::dragonbox::detail::decltype):
* Source/WTF/wtf/dragonbox/detail/util.h: Added.
* Source/WTF/wtf/dragonbox/detail/wuint.h: Added.
* Source/WTF/wtf/dragonbox/dragonbox.h: Added.
(WTF::dragonbox::to_exponential_max_string_length):
(WTF::dragonbox::to_shortest_max_string_length):
(WTF::dragonbox::max_string_length):
(WTF::dragonbox::valid_shortest_representation):
(WTF::dragonbox::count_digits_base10_with_max_17):
(WTF::dragonbox::compute_power_with_max_16):
* Source/WTF/wtf/dragonbox/dragonbox_to_chars.cpp: Added.
* Source/WTF/wtf/dragonbox/dragonbox_to_chars.h: Added.
(WTF::dragonbox::ToExponential):
(WTF::dragonbox::ToShortest):
* Source/WTF/wtf/dragonbox/ieee754_format.h: Added.
* Source/WTF/wtf/dtoa.cpp:
(WTF::numberToString):
* Source/WTF/wtf/dtoa.h:
* Source/WTF/wtf/dtoa/double-conversion.cc:
* Source/WTF/wtf/dtoa/utils.h:
(WTF::double_conversion::validShortestRepresentation):
* Tools/TestWebKitAPI/CMakeLists.txt:
* Tools/TestWebKitAPI/TestWebKitAPI.xcodeproj/project.pbxproj:
* Tools/TestWebKitAPI/Tests/WTF/DragonBoxTest.cpp: Added.
(TestWebKitAPI::DragonBoxTest::DragonBoxTest):
(TestWebKitAPI::DragonBoxTest::randomFloat):
(TestWebKitAPI::DragonBoxTest::randomDouble):
(TestWebKitAPI::DragonBoxTest::dragonBoxToString):
(TestWebKitAPI::DragonBoxTest::dragonBoxToExponential):
(TestWebKitAPI::DragonBoxTest::doubleConversionToString):
(TestWebKitAPI::DragonBoxTest::doubleConversionToExponential):
(TestWebKitAPI::TEST):

Canonical link: https://commits.webkit.org/271612@main
